ipywidgets/widgets/widget_dragdrop.py

```diff
@@ -0,0 +1,118 @@
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+"""Contains the DropWidget class"""
+from .widget import Widget, CallbackDispatcher, register, widget_serialization
+from .domwidget import DOMWidget
+from .widget_core import CoreWidget
+import json
+from traitlets import Bool, Dict, Unicode, Instance
+
+
+class DropWidget(DOMWidget, CoreWidget):
+    """Base widget for the single-child DropBox and DraggableBox widgets"""
+
+    draggable = Bool(default=False).tag(sync=True)
+    drag_data = Dict().tag(sync=True)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._drop_handlers = CallbackDispatcher()
+        self.on_msg(self._handle_dragdrop_msg)
+
+    def on_drop(self, callback, remove=False):
+        """ Register a callback to execute when an element is dropped.
+
+        The callback will be called with two arguments, the drop box
+        widget instance receiving the drop event, and the dropped element data.
+
+        Parameters
+        ----------
+        remove: bool (optional)
+            Set to true to remove the callback from the list of callbacks.
+        """
+        self._drop_handlers.register_callback(callback, remove=remove)
+
+    def drop(self, data):
+        """ Programmatically trigger a drop event.
+        This will call the callbacks registered to the  drop event.
+        """
+
+        if data.get('application/vnd.jupyter.widget-view+json'):
+            widget_mime = json.loads(data['application/vnd.jupyter.widget-view+json'])
+            data['widget'] = widget_serialization['from_json']('IPY_MODEL_' + widget_mime['model_id'])
```

Comment:
I'm thinking about this. Why is this method exposed as an API method? If the user is meant to use it from the kernel side, should we not handle the deserialization of the widget reference before calling this? This way, any kernel use could simply pass a widget reference as `widget`, and not have to worry about serializing the widget properly first.